src/dtree.ml

```diff
@@ -0,0 +1,398 @@
+              ; rhs : action
+              (** Right hand side of a rule. *)
+              ; env : term_env array
+              (** Used to pattern match variables. *)}
+
+  (** The core data, contains the rewrite rules. *)
+  type matrix = rule list
+
+  (** Type of a matrix of patterns.  Each line is a row having an attached
+      action.
+      XXX keep record? *)
+  type t = { values : matrix }
+
+  (** [pp_line o l] prints line [l] to out channel [o]. *)
+  let pp_line : line pp = List.pp Print.pp ";"
+
+  (** [pp o m] prints matrix [m] to out channel [o]. *)
+  let pp : t pp = fun oc { values = va ; _ } ->
+    let module F = Format in
+    F.fprintf oc "[|@[<v>@," ;
+    List.pp pp_line "\n  " oc (List.map (fun { lhs = l ; _ } -> l) va) ;
+    (* List.pp does not process Format "@" directives when in sep *)
+    F.fprintf oc "@.|]@,@?"
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : Terms.rule list -> t = fun rs ->
+    { values = List.map (fun r ->
+          { lhs = r.Terms.lhs ; rhs = r.Terms.rhs
+          ; env = Array.make (Bindlib.mbinder_arity r.rhs) TE_None }) rs }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> List.length m.values = 0
+
+  (** [get_col n m] retrieves column [n] of matrix [m].  There is some
+      processing because all rows do not have the same length. *)
+  let get_col : int -> t -> line = fun ind { values = valu ; _ } ->
+    let opcol = List.fold_left (fun acc { lhs = li ; _ } ->
+        List.nth_opt li ind :: acc) []
+        valu in
+    let rem = List.filter (function None -> false | Some(_) -> true) opcol in
+    List.map (function Some(e) -> e | None -> assert false) rem
+
+  (** [select m i] keeps the columns of [m] whose index are in [i]. *)
+  let select : t -> int array -> t = fun m indexes ->
+    { values = List.map (fun rul ->
+          { rul with
+            lhs = Array.fold_left (fun acc i -> List.nth rul.lhs i :: acc)
+                [] indexes }) m.values }
+
+  (** [swap p i] swaps the first column with the [i]th one. *)
+  let swap : t -> int -> t = fun pm c ->
+    { values = List.map (fun rul ->
+      { rul with lhs = List.swap_head rul.lhs c }) pm.values }
+
+  (** [cmp c d] compares columns [c] and [d] returning: +1 if [c > d], 0 if
+      [c = d] or -1 if [c < d]; where [<], [=] and [>] are defined according
+      to a heuristic.  *)
+  let cmp : line -> line -> int = fun _ _ -> 0
+
+  (** [pick_best m] returns the index of the best column of matrix [m]
+      according to a heuristic. *)
+  let pick_best : t -> int = fun _ -> 0
+
+  (** [is_pattern e t] returns whether a term [t] is considered as a pattern
+      wrt the environment [e] attached to the rule containing [t]. *)
+  let rec is_pattern : term_env array -> term -> bool = fun ar te ->
+    match te with
+    | Patt(None, _, [| |])                          -> false
```

Comment:
Why restricting the 3rd arg to [||]? Idem next line.

Reply:
Fixed in `is_pattern` in 7e04348.  However there are other places where the distinction shouldn't be done.